src/protocols/rdp/rdp.c

```diff
@@ -227,10 +227,53 @@ static BOOL rdp_freerdp_authenticate(freerdp* instance, char** username,
 
     rdpContext* context = instance->context;
     guac_client* client = ((rdp_freerdp_context*) context)->client;
-
-    /* Warn if connection is likely to fail due to lack of credentials */
-    guac_client_log(client, GUAC_LOG_INFO,
-            "Authentication requested but username or password not given");
+    guac_rdp_client* rdp_client = (guac_rdp_client*) client->data;
+    guac_rdp_settings* settings = rdp_client->settings;
+    char* params[4] = {};
+    int i = 0;
+    
+    if (settings->username == NULL || strcmp(settings->username, "") == 0) {
+        params[i] = "username";
+        rdp_client->rdp_credential_flags |= GUAC_RDP_CRED_FLAG_USERNAME;
+        i++;
+    }
+    
+    if (settings->password == NULL || strcmp(settings->password, "") == 0) {
+        params[i] = "password";
+        rdp_client->rdp_credential_flags |= GUAC_RDP_CRED_FLAG_PASSWORD;
+        i++;
+    }
+    
+    if (settings->domain == NULL || strcmp(settings->domain, "") == 0) {
+        params[i] = "domain";
+        rdp_client->rdp_credential_flags |= GUAC_RDP_CRED_FLAG_DOMAIN;
+        i++;
+    }
+    
+    /* NULL-terminate the array. */
+    params[i] = NULL;
+    
+    if (i > 0) {
+        /* Lock the client thread. */
+        pthread_mutex_lock(&(rdp_client->rdp_credential_lock));
+        
+        /* Send require params and flush socket. */
+        guac_protocol_send_required(client->socket, (const char**) params);
```

Comment:
Yeah, that makes sense.  However, question on the implementation of that related to the above comment about the `guac_socket_require_keep_alive()` call...
* Presumably this is done using the `guac_client_for_owner()` function, passing the `client` object, `guac_protocol_send_required` as the callback, and then `(const char**) params` as the data?
* Should the call to `guac_socket_require_keep_alive()` be limited to the owner, or is it safe to call that on the `client->socket` object, here?

Reply:
Okay, I think I figured this out - looks like `guac_client_for_owner()` won't work because it sends the `client->__owner` to the callback and does not send a `socket`.  So, I used `client->__owner->socket` for both the `guac_socket_require_keep_alive()` and the `guac_protocol_send_required()` call, along with the flush.

Reply:
Alrighty, a couple of issues there:

1. Nothing should be referencing the private portions of libguac structures except libguac. To safely access the owner of a connection, you really do need to use `guac_client_for_owner()`.

   It should work just fine. In any case that you would be tempted to reference `client->__owner->whatever`, you can instead do `user->whatever` for the user passed to the callback.

2. You probably actually want to use the broadcast socket for `guac_socket_require_keep_alive()`.

   The utility of that function is that it ensures the other side of the connection doesn't drop out while waiting for data from the server. It is true that we probably only want to request `argv` from the owner, but if the connection is shared, we don't want _any_ of those connections to drop out. A user sharing the connection shouldn't get disconnected just because the owner of the connection is taking their time to enter their credentials.

Reply:
> Using guac_client_for_owner() should work just fine whenever you need a guac_user. In any case that you would be tempted to reference client->__owner->whatever, you can instead do user->whatever for the user passed to the callback.

So, should the `guac_protocol_send_required()` function be re-worked to take `guac_user` instead of a `guac_socket`?  Or should I implement some intermediate function in `user.c` that uses this `guac_protocol_send_required()` function and just passes through the `guac_user->socket`?

Reply:
> So, should the `guac_protocol_send_required()` function be re-worked to take `guac_user` instead of a `guac_socket`?

No, I think the `guac_protocol_send_*()` family of functions should remain unaware of the concerns of `guac_user`, `guac_client`, etc. They are intended as low-level functions for sending instructions over a `guac_socket`.

> Or should I implement some intermediate function in `user.c` that uses this `guac_protocol_send_required()` function and just passes through the `guac_user->socket`?

This would be the way, yes.

As this would be a common need, if you're looking to avoid having to do this everywhere, I think it could make sense to provide a convenience function like `guac_client_owner_send_required()` or similar.

Reply:
Okay, shot has been taken at an implementation of `guac_client_owner_send_required()`.  Guessing it'll need some tweaking, but it's a start.